### PR TITLE
feat: PER-7348 add waitForReady() call before serialize()

### PR DESCRIPTION
**Pull request by Shivanshu-07**
## Summary

Adds Pattern C (in-browser) readiness gate ahead of `PercyDOM.serialize()` per PER-7348.

Pairs with: https://github.com/percy/cli/pull/2184

### Behavior

- Before serialize, the SDK calls `PercyDOM.waitForReady(readinessConfig)` — async — and awaits the promise.
- `typeof PercyDOM.waitForReady === 'function'` guard preserves backward compatibility with older CLI builds that don't ship the readiness API.
- `readiness.preset === 'disabled'` short-circuits — `waitForReady` is not called.
- If `waitForReady` rejects, the SDK logs at debug level and proceeds to `serialize()` so snapshot capture is never blocked.
- Diagnostics returned by `waitForReady` are attached to `domSnapshot.readiness_diagnostics` so the CLI can log timing and pass/fail.
- Readiness config resolves from per-snapshot `options.readiness` first, then `utils.percy.config.snapshot.readiness` from `.percy.yml`.

`PercyDOM.serialize()` itself is unchanged — still synchronous, called with the same options as today.

### SDK-visible contract

```js
// New — readiness step (skipped on old CLI / disabled preset)
if (preset !== 'disabled' && typeof window.PercyDOM?.waitForReady === 'function') {
  diagnostics = await window.PercyDOM.waitForReady(config);
}

// Unchanged
let domSnapshot = window.PercyDOM.serialize({ ...options });
if (diagnostics) domSnapshot.readiness_diagnostics = diagnostics;
```

## Test plan

QUnit acceptance tests added to `tests/acceptance/index-test.js` under a `readiness gate (PER-7348)` module:

- [x] `waitForReady` is called before `serialize` when CLI exposes it
- [x] Backward compat — only `serialize` runs when `waitForReady` is absent
- [x] `readiness.preset: disabled` skips `waitForReady`
- [x] When `waitForReady` rejects, `serialize` still runs
- [x] Diagnostics returned by `waitForReady` are attached to `domSnapshot.readiness_diagnostics`

> ⚠️ Unit tests written but not executed locally (missing ember-cli toolchain in the fan-out environment). CI will exercise them.

Smoke test against `example-percy-ember` was not run as part of this fan-out.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: addon-test-support/@percy/ember/index.js [hunk 1]
```diff
@@ -76,6 +76,18 @@ export default async function percySnapshot(name, {
 
     addPseudoClassEnabledELements(options);
 
+    // Readiness gate — runs before serialize when CLI supports it (PER-7348).
+    // Uses typeof guard for backward compat with older CLI that lacks waitForReady.
+    let readinessDiagnostics;
+    const readinessConfig = options.readiness || utils.percy?.config?.snapshot?.readiness || {};
+    if (readinessConfig.preset !== 'disabled' && typeof window.PercyDOM?.waitForReady === 'function') {
```
Comment (pranavz28) [inline addon-test-support/@percy/ember/index.js]: 🔴 **P1 — Blocks merge: `await waitForReady` re-opens the Ember run loop, then sync `serialize()` runs against a DOM frame the user never sees painted.**

The CLI's readiness implementation (percy/cli#2184) installs `MutationObserver`, two `PerformanceObserver`s (`resource`, `longtask`), `setTimeout`, polling `setInterval` (fallback), and `rAF`/`rIC`. Every one of those ticks a microtask/macrotask in the Ember test environment.

The sequence inside a QUnit acceptance test:
1. Test author does `await settled()` → Backburner quiescent, DOM matches the picture the user wants captured.
2. Test author does `await percySnapshot(assert)`.
3. We hit this line, `await window.PercyDOM.waitForReady(...)` runs for up to 10s (`balanced` preset default), during which the observers and rAF callbacks fire as microtasks. Each one is an opportunity for Backburner to pick back up — e.g. a mutation observer firing flips a tracked property which schedules a glimmer rerender on the next frame.
4. `waitForReady` resolves on this same frame. We immediately call `window.PercyDOM.serialize(...)` synchronously on line 92.
5. The captured DOM may be mid-rerender — half-applied glimmer output that would never have been painted in a real browser.

This is exactly the "manufactured flicker" anti-pattern visual regression tooling exists to catch. Snapshots will be intermittently wrong in ways that look like product bugs.

**Fix (≈3 lines):**
```js
import { settled } from '@ember/test-helpers';
// ...
// after the readiness try/catch block, before serialize:
if (typeof settled === 'function') await settled();
```

`settled()` already waits for the runloop, routing, AJAX, waiters, and `rAF` — exactly the surface readiness can perturb.

File: addon-test-support/@percy/ember/index.js [hunk 2]
```diff
@@ -76,6 +76,18 @@ export default async function percySnapshot(name, {
 
     addPseudoClassEnabledELements(options);
 
+    // Readiness gate — runs before serialize when CLI supports it (PER-7348).
+    // Uses typeof guard for backward compat with older CLI that lacks waitForReady.
+    let readinessDiagnostics;
+    const readinessConfig = options.readiness || utils.percy?.config?.snapshot?.readiness || {};
```
Comment (pranavz28) [inline addon-test-support/@percy/ember/index.js]: 🟡 **P2: `||` short-circuit prevents explicit-falsy override of `.percy.yml` config.**

The PR description says "`options.readiness` first, then `utils.percy.config.snapshot.readiness`" — but `||` doesn't honor explicit falsy overrides. A caller passing `{ readiness: null }` or `{ readiness: false }` to opt out at the per-snapshot level falls back to the global config and still runs the gate, contradicting their intent.

**Fix:**
```js
const readinessConfig = options.readiness ?? utils.percy?.config?.snapshot?.readiness ?? {};
```

Add a test that locks the precedence direction (mirroring the existing pattern at lines 143-160 for `pseudoClassEnabledElements`). That test would also have caught this issue.

File: addon-test-support/@percy/ember/index.js [hunk 3]
```diff
@@ -76,6 +76,18 @@ export default async function percySnapshot(name, {
...
+    if (readinessConfig.preset !== 'disabled' && typeof window.PercyDOM?.waitForReady === 'function') {
+      try {
+        readinessDiagnostics = await window.PercyDOM.waitForReady(readinessConfig);
```
Comment (pranavz28) [inline addon-test-support/@percy/ember/index.js]: 🟡 **P2: `window.PercyDOM` rebinding race when a caller forgets `await`.**

This is a real footgun for in-browser SDKs. Consider:
```js
test('something', async function(assert) {
  percySnapshot(assert);   // forgot to await
  await click('.x');
  await percySnapshot(assert, { name: 'after' });
});
```
The first call's `await waitForReady` is suspended. The second call's `eval(fetchPercyDOM())` at line 74 may re-install `window.PercyDOM`, or QUnit teardown nulls it. The first call then resumes and reads `window.PercyDOM.serialize` (line 92) against a different (or absent) PercyDOM build. Diagnostics from build A get attached to a serialization from build B, and the user has no visible signal of the cross-build mix-up.

The new tests' `beforeEach`/`afterEach` (lines 164-172) hide this — they `await` percySnapshot fully each time, so the rebinding never overlaps the suspended call.

**Fix (one line):** capture a stable reference once, use it throughout:
```js
const PercyDOM = window.PercyDOM;  // after the fetchPercyDOM block
// then use PercyDOM.waitForReady, PercyDOM.serialize below
```

File: addon-test-support/@percy/ember/index.js [hunk 4]
```diff
@@ -76,6 +76,18 @@ export default async function percySnapshot(name, {
...
+      } catch (e) {
+        log.debug(`waitForReady failed, proceeding to serialize: ${e?.message || e}`);
```
Comment (pranavz28) [inline addon-test-support/@percy/ember/index.js]: 🟡 **P2: `log.debug` will be invisible to customers.**

`@percy/sdk-utils`' logger default is `info`. If a customer's readiness config systematically fails (bad selector, unsupported preset, CLI bug), every snapshot in their suite silently bypasses readiness and captures a too-early DOM — and they get no warning louder than a debug log they're not capturing.

The CLI PR's spec for Pattern C is `.catch(() => {})` (completely silent), so this SDK is already more informative — just at the wrong level.

**Fix:**
```js
} catch (e) {
  readinessDiagnostics = { error: e?.message || String(e), proceeded: true };
  log.warn(`waitForReady failed, proceeding to serialize: ${e?.message || e}`);
}
```

Upgrade to `warn`, and stamp the failure into `readiness_diagnostics` so the CLI can render it next to the snapshot (otherwise rejections are indistinguishable from "old CLI, gate never ran"). Also fixes the `e?.message || e` empty-message edge case for free.

File: addon-test-support/@percy/ember/index.js [hunk 5]
```diff
@@ -84,6 +96,11 @@ export default async function percySnapshot(name, {
       ...options
     });
 
+    // Attach readiness diagnostics so the CLI can log timing and pass/fail
+    if (readinessDiagnostics) {
+      domSnapshot.readiness_diagnostics = readinessDiagnostics;
```
Comment (pranavz28) [inline addon-test-support/@percy/ember/index.js]: 🟡 **P2: Diagnostics shape not validated before assignment → can take down the entire snapshot it was supposed to instrument.**

This assigns whatever `waitForReady` returned directly onto `domSnapshot.readiness_diagnostics`. `utils.postSnapshot` (line 110) eventually `JSON.stringify`'s the payload. Per CLI #2184 the shape is well-defined today (`{ passed, timed_out, preset, checks, [error] }`), but a future regression returning a circular ref, function, BigInt, or DOM node throws inside `postSnapshot` → caught by the outer try/catch at line 113 → `Could not take DOM snapshot` is logged and the snapshot never posts.

A diagnostics field designed to *instrument* the snapshot then takes down the snapshot.

**Fix:**
```js
if (readinessDiagnostics !== undefined) {
  try {
    JSON.stringify(readinessDiagnostics);
    domSnapshot.readiness_diagnostics = readinessDiagnostics;
  } catch (e) {
    log.warn(`dropping unserializable readiness diagnostics: ${e?.message}`);
  }
}
```

Note: `if (readinessDiagnostics)` also drops legitimate falsy values (e.g. resolve-with-`null` to indicate "gate ran, no diagnostics"). `!== undefined` is more precise.

File: addon-test-support/@percy/ember/index.js [hunk 6]
```diff
@@ -76,6 +76,18 @@ export default async function percySnapshot(name, {
 
     addPseudoClassEnabledELements(options);
 
+    // Readiness gate — runs before serialize when CLI supports it (PER-7348).
+    // Uses typeof guard for backward compat with older CLI that lacks waitForReady.
```
Comment (pranavz28) [inline addon-test-support/@percy/ember/index.js]: 🔵 **P3 nit: line 79 restates structure visible from the code (the gate sits right before `serialize`; the `typeof ... === 'function'` check obviously means "when CLI supports it"). Line 80 is the genuine non-obvious WHY and is worth keeping.**

Suggest collapsing both lines to:
```js
// Backward-compat: older CLI bundles lack waitForReady (PER-7348).
```

The comment at line 99 (`Attach readiness diagnostics so the CLI can log timing and pass/fail`) is borderline but worth keeping — it explains the consumer's purpose, which isn't visible from this file alone.

File: tests/acceptance/index-test.js [hunk 7]
```diff
@@ -160,6 +160,102 @@ module('percySnapshot', hooks => {
...
+      const calls = [];
+      window.PercyDOM = {
+        waitForReady: cfg => { calls.push(['waitForReady', cfg]); return Promise.resolve(); },
+        serialize: opts => { calls.push(['serialize', opts]); return { html: '<html></html>' }; }
+      };
+
+      await percySnapshot('readiness-happy-path');
+
+      const order = calls.map(([n]) => n);
+      assert.ok(order.indexOf('waitForReady') < order.indexOf('serialize'),
```
Comment (pranavz28) [inline tests/acceptance/index-test.js]: 🔵 **P3: Order assertion is too loose.**

`assert.ok(order.indexOf('waitForReady') < order.indexOf('serialize'))` only checks that the **first** waitForReady precedes the first serialize. It also passes for `['waitForReady', 'waitForReady', 'serialize']` or `['waitForReady', 'serialize', 'waitForReady']` — double-invocation regressions slip past silently.

Tighten to match the style used in every other test in this module (lines 197, 210, 223, 238):
```js
assert.deepEqual(order, ['waitForReady', 'serialize'],
  'waitForReady is called exactly once before serialize');
```

File: tests/acceptance/index-test.js [hunk 8]
```diff
@@ -160,6 +160,102 @@ module('percySnapshot', hooks => {
...
+      assert.deepEqual(calls.map(([n]) => n), ['waitForReady', 'serialize'],
+        'serialize still runs after waitForReady rejection');
+    });
+
+    test('proceeds to serialize when waitForReady rejects with a non-Error', async assert => {
+      // Covers the `e?.message || e` second branch: rejection value has
+      // no `.message`, so logging falls through to stringifying e itself.
+      const calls = [];
+      window.PercyDOM = {
+        waitForReady: () => { calls.push(['waitForReady']); return Promise.reject('plain-string-rejection'); },
```
Comment (pranavz28) [inline tests/acceptance/index-test.js]: 🔵 **P3: This test was added in commit 6dfd376 specifically to cover the `|| e` branch of `e?.message || e` (per the inline comment at line 228). That's a line-coverage-shaped test, not a behavior test — it asserts only the same call order as the Error-rejection test at line 223 and never inspects the logged value.**

If someone replaced `e?.message || e` with `String(e)` or `e instanceof Error ? e.message : e`, this test would still pass even though the *thing it exists to cover* changed.

Two paths forward:
1. Make it behavioral — assert that the logger captured `'plain-string-rejection'`.
2. Replace the idiom on `addon-test-support/@percy/ember/index.js:87` with `e instanceof Error ? e.message : String(e)` — clearer intent, no empty-message edge case, and the line-coverage rationale for this test disappears (the test for the rejection branch at line 223 is sufficient).

If you fix the wider P2 finding on line 87 (upgrade to `log.warn` and stamp failure into diagnostics), this test becomes obsolete anyway — replace it with one that asserts `domSnapshot.readiness_diagnostics.error === 'plain-string-rejection'`.

**Review by pranavz28 — COMMENTED** [review]
## Code review summary

Cross-referenced against the merged CLI PR percy/cli#2184, which gives us the contract for `waitForReady` (always resolves within `min(timeout_ms, max_timeout_ms)`; never hangs; returns `{ passed, timed_out, preset, checks, [error] }`; `readiness_diagnostics` is the intentional snake_case wire field; default preset is `balanced` when no config provided).

**1 blocker, 4 should-fix, 4 nice-to-have.** Detailed findings inline.

### 🔴 P1 — Blocks merge
- **Ember run loop re-opens during `await waitForReady`** — readiness checks install MutationObserver, PerformanceObserver, rAF/rIC, all of which tick Backburner. By the time `serialize()` runs synchronously, the DOM may be mid-render. Fix: `await settled()` from `@ember/test-helpers` after readiness, before serialize.

### 🟡 P2 — Should fix in this PR
- `window.PercyDOM` rebinding race when a caller forgets to `await percySnapshot()`
- `||` short-circuit prevents `{ readiness: null }` from overriding `.percy.yml` config
- `log.debug` swallows readiness failures invisibly (default log level is info)
- Diagnostics passed unchecked into `postSnapshot` → `JSON.stringify` can throw and lose the snapshot

### 🔵 P3 — Nice-to-have
- Sloppy order assertion (passes for `['waitForReady','serialize','waitForReady']`)
- Non-Error rejection test is line-coverage-shaped — asserts call order, not the logged value it was added to cover
- Restating-WHAT comment
- Testing coverage gaps: no test for the `utils.percy.config.snapshot.readiness` fallback path, no shape assertion on the `cfg` passed to `waitForReady`, no precedence test between per-call and config-level readiness, no test pinning behavior for `{ timeout: 5000 }` (readiness keys without preset)

### Non-findings / cross-checked away
- `readiness_diagnostics` snake_case is confirmed by CLI schema — do not rename
- Default-ON-when-no-preset is confirmed product design (`balanced` is intentional)
- `'disabled'` exact-string match is consistent CLI↔SDK — not a divergence bug
- No SDK-side timeout — acceptable, CLI guarantees resolution
- Agent-native parity: PASS (library API, no UI surface)

No prior solutions found in `docs/solutions/` for `waitForReady` / PER-7348 / readiness gate — first SDK to adopt Pattern C, worth a writeup after merge so the other in-browser SDKs (Cypress) can copy the corrected pattern.

**Review by rishigupta1599 — APPROVED** [review]
